[nlqhung130403] iP

nlqhung130403 opened the pull request:
# Ferb

> “Your mind is for having ideas, not holding them.” – David Allen [(source)](https://dansilvestre.com/productivity-quotes/)

Ferb helps you stay organized and never forget tasks again. It's:

- **Command-line based**
- **User-friendly**
- **Blazingly FAST to use**

### Getting Started:
1. Download it from [here](https://github.com/nlqhung130403/ip).
2. Run the program.
3. Add your tasks effortlessly.
4. Let Ferb handle the rest for you! 🚀

And the best part? **It’s completely FREE!**

### Features:
- ✅ Task management
- ⬜ Deadline tracking (coming soon)
- ⬜ Smart reminders (coming soon)

---

If you're a Java developer, you can also use Ferb to sharpen your Java skills. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(FerbApp.class, args);
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

nlqhung130403 committed:
Change names and greetings

nlqhung130403 committed:
implement echo

nlqhung130403 committed:
implement Add, List

nlqhung130403 committed:
Implement Mark As Done

nlqhung130403 committed:
Implement Todo, Deadline and Event

nlqhung130403 committed:
Handle exception

nlqhung130403 committed:
Add delete and update level-5 task

nlqhung130403 committed:
Perform text-UI testing

nlqhung130403 committed:
Read and Write to file

nlqhung130403 committed:
Merge branch 'branch-Level-7'

nlqhung130403 committed:
Implement date feature

nlqhung130403 committed:
Merge branch 'branch-Level-8'

TobyCyan commented on src/main/java/FerbFileHandler.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.List;
+
+public class FerbFileHandler {
+    private String filepath;
```
Should this part be filePath instead?
```suggestion
    private String filePath;
```

TobyCyan commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,45 @@
+        return this.displayDone() + this.taskDescription();
+    }
+
+    public String toSave() {
```
I like how you gave short and intuitive names to these methods❤️

TobyCyan commented on src/main/java/Ferb.java:
```diff
@@ -0,0 +1,137 @@
+            String command = scanner.nextLine();
+            try {
+                Task task;
+                if (command.equals("bye")) {
```
Hey, would like to know any reason why you used if else statements instead of switch case?

TobyCyan commented on src/main/java/Ferb.java:
```diff
@@ -0,0 +1,137 @@
+        String[] tasks = content.split("\\R");
+
+        for (String task : tasks) {
+            String[] fields = task.split("\\|"); //escape special character
```
I like how you used plural forms for naming collections, nice 👍

TobyCyan commented on src/main/java/Ferb.java:
```diff
@@ -0,0 +1,137 @@
+        System.out.println("Here are the tasks in your list:");
+        for (int i = 1; i <= tasks.size(); i++) {
+            Task task = tasks.get(i-1);
+            System.out.println(INDENT + i + "." + task.toString());
```
I noticed there are a lot of print statements with the same format, perhaps you can create a method for it?

TobyCyan commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,36 @@
+    public String toString() {
+        return "[E]" + this.displayDone() + this.taskDescription()
+                + " (from: " + Event.displayDate(this.startDate) + " to: "
+                + Event.displayDate(this.endDate) + ")";
```
I like how you did the line breaks 👏. And I noticed you did the same for some other parts as well!

TobyCyan commented on src/main/java/Ferb.java:
```diff
@@ -0,0 +1,137 @@
+            String type = fields[0];
+            boolean isDone = fields[1].equals("1") ? true : false;
+            String description = fields[2];
+            if (type.equals("T")) {
```
Perhaps a switch case here would be more readable?

TobyCyan commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,45 @@
+public class Task {
+    private boolean isDone;
```
I like how you named boolean variables like questions 👏

TobyCyan reviewed (commented):
Overall, I find the code to follow the coding standards very well, except for one part where I have noted a minor coding standard violation. I really like how well the codebase is structured, methods have short and very intuitive naming, and things are very easy to trace. I have noted a part where there is very long series of if-else statements. Consider using switch case for better readability if that makes sense.

Noob-No-1 commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,33 @@
+        super(isDone, description);
+        this.deadline = LocalDate.parse(deadline);
+    }
+
+    public String displayDeadline() {
```
Maybe adding a documentation for the public non-trivial method can help other collaborators to understand your method faster without navigating to this specific section of the code?

Noob-No-1 commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,36 @@
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+
```
Maybe can consider to add documentation for classes?

Noob-No-1 commented on src/main/java/Ferb.java:
```diff
@@ -0,0 +1,137 @@
+            } else {
+                String from = fields[3];
+                String to = fields[4];
+                Ferb.tasks.add(new Event(isDone, description, from, to));
```
I noticed you maybe assumed there will not be invalid input (i.e. any character outside of T, D and E), this may lead to runtime error when unexpected cases are encountered. Maybe can do a try catch block with an additional default switch case to handle invalid input format?

Noob-No-1 commented on src/main/java/Ferb.java:
```diff
@@ -0,0 +1,137 @@
+                    throw new FerbException();
+                }
+                Ferb.add(task);
+            } catch (StringIndexOutOfBoundsException e) {
+                System.out.println(INDENT + "Wrong command format! Please double check.");
+            } catch (IndexOutOfBoundsException e) {
+                System.out.println(INDENT + "Invalid Index! Please try again.");
+            } catch (RuntimeException e) {
+                System.out.println(INDENT + "Wrong command format! Please try again.");
+            } catch (FerbException e) {
+                System.out.println(INDENT + "Sorry! Command not supported!");
```
Good to see that you are anticipating exceptions and handling them well! Maybe can use switch statements for neater code and also add default case to handle invalid input format?

Noob-No-1 commented on src/main/java/FerbException.java:
```diff
@@ -0,0 +1,6 @@
+public class FerbException extends Exception{
+
+    public FerbException() {
```
Maybe it will be good to pass the message as constructor argument to print error message on the console? It will come in handy in the debug process.

nlqhung130403 committed:
Implement More OOP feature

nlqhung130403 committed:
Merge branch 'A-MoreOOP'

nlqhung130403 committed:
Organise files into packages

nlqhung130403 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

nlqhung130403 committed:
Add gradle

nlqhung130403 committed:
Implement find feature

nlqhung130403 committed:
Merge branch 'branch-Level-9'

nlqhung130403 committed:
Test multiple classes

nlqhung130403 committed:
Add JavaDocs

nlqhung130403 committed:
Modify code to follow coding standard

nlqhung130403 committed:
Merge branch 'branch-A-JavaDoc'

nlqhung130403 committed:
Merge branch-A-CodingStandard

nlqhung130403 committed:
Implement GUI using JavaFX

nlqhung130403 committed:
Merge branch 'branch-Level-10'

nlqhung130403 committed:
Improve code quality of parser and commands

nlqhung130403 committed:
Improve code quality of TaskList

nlqhung130403 committed:
Merge branch 'branch-A-CodeQuality'

nlqhung130403 committed:
Add assertions for assumptions

nlqhung130403 committed:
Merge branch 'branch-A-Assertions'

nlqhung130403 committed:
Add more assertions

nlqhung130403 committed:
Merge pull request #2 from nlqhung130403/branch-A-Assertions

Add more assertions

nlqhung130403 committed:
Implement C-Sort using merge sort

nlqhung130403 committed:
Implement C-Sort

nlqhung130403 committed:
Implement Sort command

Allow users to sort TaskList using sort command based on either task
descriptions or task dates (if applicable).

Use quicksort to implement sortDate() as it is an efficient and in-place
sorting algorithm.

nlqhung130403 committed:
Add javadocs for public methods

nlqhung130403 committed:
Modify unit tests following new implementation

nlqhung130403 committed:
Add testunits for TaskList

nlqhung130403 committed:
Add more tasks to thoroughly test TaskList sort methods

nlqhung130403 committed:
Add User Guide and Ui image

nlqhung130403 committed:
Delete an extra line in Readme.md

nlqhung130403 committed:
Update Readme to include installation guide

nlqhung130403 committed:
Fix jar file's link
